Minimal unwinding information (DWARF CFI) checker

amsen20 opened the pull request:
This PR adds a minimal version of `UnwindInfoChecker` described in [here](https://discourse.llvm.org/t/rfc-dwarf-cfi-validation/86936). 

This implementation looks into the modified registers by each instruction and checks:
- If one of them is the CFA register, and it informs the user if the CFA data is not modified as well.
- If one of them is used in another register's unwinding rule, it informs the user if the unwind info is not modified after this instruction.

This implementation does not support DWARF expressions and treats them as unknown unwinding rules.

Commit:
Forward CFI directives alongside their associated instruction to a simple CFI analysis

Commit:
Create a new CFI analysis for each dwarf separately

Commit:
Separate the logic and display

Commit:
Compile llvm mc depending on bolt

Commit:
Extract semantic info on MCInst from bolt

Commit:
Typo in evaluateStackOffsetExpr doc

Commit:
Add a DS to keep track of each register CFI state

Commit:
Check if the CFA change is correct or not

Commit:
Add the CFI validation as a option to llvm-mc tool

Commit:
Provide the directives as an arrayref instead of index range

Commit:
Remove TODOs

Commit:
Wrap all the MCPB info needed into a ExtendedMCInstrAnalysis class

Commit:
Use extended MCAnalysis instead of MCplus (directly)

Commit:
Make the CFA checker to first generate, if not matched check the Read/Write conditions

Commit:
Check register states as well

Commit:
Process prologue as well

Commit:
Ignore flag registers as well

Commit:
Enable dump

Commit:
Enable getting flag register through extended MCInstrAnalysis

Commit:
Move the isolated MC semantic methods to ExtendedMCAnalysis

Commit:
Remove unused includes

Commit:
Test cfi validator on following cases:
- Changing the register and not informing it using cfi directives
- Changing the CFA and report it wrong
- Load a register from another register stored location

Commit:
Prune the CFA diff checker to only check read/write

Commit:
Move the CFI analysis headers to MC

Commit:
Move all the CFI analysis implementations to lib

Commit:
Remove stack pointer method

Commit:
Comment out Ignoring flags at the beginning CFI analysis

Commit:
Move the necessary MCStreamer interface methods to CFIAnalysisMCStreamer.h

Commit:
Make the update method const

Commit:
Move CFI analysis out of MC layer

Commit:
Write a CFI directive to DWARF operation converter

Commit:
Use UnwindTable instead of CFI State:

- Change unwind table (by making small things public) to be usable from CFIAnalysis
- Remove CFIState internal DS and use UnwindTable instead
- Add a MCCFIInstruction -> CFIProgram::Instruction converter
- Refactor CFIAnalysis to use UnwindTable

Commit:
Make CFI analysis available both EH and Debug frames

Commit:
Remove ExtendedMCA

Commit:
Use UnwindRow as state and record the UnwindHistory

Commit:
Make CFA access easier

Commit:
Rename CFI to Unwind info analysis

Commit:
Remove MCInstrAnalysis from UnwindInfo analysis

Commit:
Remove redundant definitions and fields from CFIAnalysis streamer

Commit:
Rename CFIAnalysisStreamer to FunctionUnitStreamer

Commit:
Test function unit streamer functionalities

Commit:
Fix the CFIProgram usage temporary

Commit:
Minimize check reg diff arguments

Commit:
Move `FunctionUnitStreamer` analysis calls to a different class called `FunctionUnitAnalyzer`.

Define `FunctionUnitUnwindInfoAnalyzer` that executes the Analysis during the stream.

The main reasons for doing this is:
- This is compatible with other use cases of `MCStreamer` implementers (e.g. `MCAsmStreamer` and `MCAssembler`).
- This approach result in on the fly analysis, if it was implemented like `RecordStreamer`, then all the instruction of the whole program have to be copied and be stored in the memory.

Commit:
Rename the library to UnwindInfoChecker

Commit:
Fix the header guards

Commit:
Add file headers

Commit:
Resolve warnings

Commit:
Refactor UnwindInfoAnalysis

Commit:
Move tests to UnwindInfoChecker folder

Commit:
Rename UnwindHistory to UnwindState

Commit:
Make different methods to add instruction public

Commit:
Revert bolt change

Commit:
Revert bolt changes

Commit:
Change UnwindTable to RowContainer

Commit:
Refactor UnwindInfoState

Commit:
Revert unnecessary changes to llvm-mc

amsen20 commented:
Please help me in:
- Breaking this implementation in re-viewable parts
- Testing each part of the implementation properly

It's my first non-mechanical PR, so any suggestion, even little is appreciated.

Commit:
Lint

amsen20 commented:
I see the following plan fit for splitting the implementation:
- Edits to `DebugInfo` can be their own PR, they are minor as well.
- `FunctionUnitStreamer` and `FunctionUnitAnalyzer` can be moved to `llvm/MC/`, like other `MCStreamer` implementers, and they can be tested with unit testing separately.
- The remaining can stay in this PR, depending on last independent PRs.

Commit:
Pass the tests

ilovepi reviewed (commented):
This is quite a large patch. It may make sense to split out parts that aren't required for complete functionality into their own patches.

For example, making `addInstruction()` public in DWARFCFIProgram.h, can probably be its own patch. I think there are several places of where you're just slightly refactoring the existing code.

As for the rest ... Its rather large, and reviewing incremental changes as you add functionality would make it significantly easier to provide high quality reviews.

Commit:
Fix weird end file formatting issues

Commit:
Apply suggestions

Commit:
Make class names more specific

Commit:
Change file names based on their class names

Commit:
Rename tests folder name

amsen20 commented on llvm/include/llvm/UnwindInfoChecker/FunctionUnitAnalyzer.h:
```diff
@@ -0,0 +1,49 @@
+//===----------------------------------------------------------------------===//
```
Sorry, the commit that you are reviewing was in the middle of some refactoring I've been doing, I finished them right now.
Class names, file names, matches now.

Commit:
Remove uneccasary optional

Commit:
Summarize instantiation

Commit:
Either add an instruction or crash

amsen20 commented on llvm/lib/UnwindInfoChecker/UnwindInfoState.cpp:
```diff
@@ -0,0 +1,144 @@
+}
+
+std::optional<dwarf::CFIProgram>
+DWARFCFIState::convert(MCCFIInstruction Directive) {
```
Fixed. 

Only two operations don't add an instruction:
- Expressions are ignored for now.
- OpLabel should be implemented, for now, this instruction should not be reached by the tool.

Commit:
Document the process of updating CFI state

Commit:
Document why register information should not vanish or appear during analysis

Commit:
Change unnessary assert to warning

Commit:
Assert ref reg when using it

amsen20 commented on llvm/lib/UnwindInfoChecker/UnwindInfoAnalysis.cpp:
```diff
@@ -0,0 +1,299 @@
+  switch (PrevLoc.getLocation()) {
+  case dwarf::UnwindLocation::Same:
+  case dwarf::UnwindLocation::RegPlusOffset:
+    if (Writes.count(MaybePrevRefReg.value())) {
```
When the unwind location is `Same` or `RegPlusOffset`, there is always a reference register, so no need for checking. But I added an assert.

Commit:
Change register naming format from %{reg} to register ${regName}

amsen20 commented on llvm/lib/UnwindInfoChecker/UnwindInfoAnalysis.cpp:
```diff
@@ -0,0 +1,299 @@
+    if (PrevLoc.getLocation() == NextLoc.getLocation()) {
+      Context->reportWarning(
+          Inst.getLoc(),
+          formatv("unknown change happened to %{0} unwinding rule values",
```
Fixed, changed it to register ${regName}.

Commit:
Make the tests more restrict

Commit:
File check test as well

Commit:
Fix misspells

amsen20 commented on llvm/test/DWARFCFIChecker/X86/spill-two-reg.s:
```diff
@@ -10,28 +11,34 @@ _start:
         .cfi_same_value %rsi
 
         pushq   %rbp
+        # CHECK: warning: unknown change happened to register RBP unwinding rule structure
```
Fixed.

Commit:
Move the empty implementations to header

amsen20 commented on llvm/include/llvm/DWARFCFIChecker/DWARFCFIFunctionFrameReceiver.h:
```diff
@@ -0,0 +1,49 @@
+
+  MCContext &getContext() const { return Context; }
+
+  virtual void startFunctionUnit(bool IsEH,
```
Fixed

amsen20 commented on llvm/lib/DWARFCFIChecker/DWARFCFIAnalysis.cpp:
```diff
@@ -0,0 +1,306 @@
+  auto PrevRow = MaybePrevRow.value();
+
+  for (auto &&Directive : Directives)
+    State.update(Directive);
```
I changed it to be a `std::vector<dwarf::UnwindRow*>`. It's not cache-efficient but I chose it because I need to:
- Append new rows, so I cannot use `UnwindTable` or a static array.
- Be able to pass them around without copying them.
- Mark them as `const` so they are not mutable after being appended.

Also, unlike the previous iterator design, I don't want other rows to be accessible from each other.

Commit:
Use vector of pointers instead of vector of UnwindRow objects to avoid memory safety bugs that can happen due to vector resizing.

Commit:
Document each class resposibility and rules

amsen20 commented on llvm/include/llvm/DWARFCFIChecker/DWARFCFIState.h:
```diff
@@ -0,0 +1,42 @@
+//===----------------------------------------------------------------------===//
+///
+/// \file
+/// This file declares DWARFCFIState class.
```
Added documentation

Commit:
Remove mistaked documentation

amsen20 commented on llvm/include/llvm/DWARFCFIChecker/DWARFCFIFunctionFrameStreamer.h:
```diff
@@ -0,0 +1,71 @@
+//===----------------------------------------------------------------------===//
+///
+/// \file
+/// This file declares CFIFunctionFrameStreamer class.
```
Added documentation

amsen20 commented on llvm/include/llvm/DWARFCFIChecker/DWARFCFIFunctionFrameReceiver.h:
```diff
@@ -0,0 +1,49 @@
+//===----------------------------------------------------------------------===//
+///
+/// \file
+/// This file declares CFIFunctionFrameReceiver class.
```
Added documentation

Sterling-Augustine replied:
What motivates the abstract base class here? it feels a little yagni, and adds complexity that may not be realistically needed. But perhaps there are plans for multiple implementations later?

amsen20 replied:
Two reasons motivate the existence of this abstract class:
1. The main reason for developing this class is to enable `DWARFCFIFunctionFrameStreamer` to be moved to any other library (e.g., `DebugInfo/DWARF` or `MC`).
2. There might be other uses for a stream of FDEs in instruction per instruction flow; a possible use case that I had in mind was a dummy printer implementation that only prints the instructions and their directives in order.

If these reasons are not enough to have this abstract class, I can just remove it.

amsen20 commented on llvm/include/llvm/DWARFCFIChecker/DWARFCFIFunctionFrameAnalyzer.h:
```diff
@@ -0,0 +1,42 @@
+//===----------------------------------------------------------------------===//
+///
+/// \file
+/// This file declares CFIFunctionFrameAnalyzer class.
```
Added documentation

amsen20 commented on llvm/lib/UnwindInfoChecker/UnwindInfoAnalysis.cpp:
```diff
@@ -0,0 +1,299 @@
+    Context->reportWarning(
+        Inst.getLoc(),
+        formatv("unknown change happened to %{0} unwinding rule structure",
+                RegName));
```
I added this to the `DWARFCFIAnalysis` description. It's a bit modified and not synced with the implementation. I will change the implementation accordingly and then mentioned each case in the code.

statham-arm commented on llvm/include/llvm/DWARFCFIChecker/DWARFCFIFunctionFrameStreamer.h:
```diff
@@ -24,17 +24,22 @@
-        Analyzer(std::move(Analyzer)), LastDirectiveIndex(0) {
-    assert(this->Analyzer && "Analyzer should not be null");
+        Receiver(std::move(Receiver)), LastDirectiveIndex(0) {
+    assert(this->Receiver && "Analyzer should not be null");
```
I like the rename of `Analyzer` to the more purpose-neutral `Receiver`, but you left one unchanged in the assert failure message here :slightly_smiling_face:

amsen20 replied:
Yes, the main reason for the previous `iterator` design was that when reporting the error to the user, we want to be able to refer to the exact directive that caused this error. To do that, the only way is to iterate the table from the last row to the current row, one by one, find the row that applies the change, and report the error to the corresponding directive.

But for now, in this version, errors are reported to the instruction. I can:
- Try to figure out a way to report the error to the directive: The logic is not very simple. To do so, I need to add a row <-> directive mapping. Additionally, between the last row and the current row, a value may change multiple times, which means we must choose a policy to determine which changes correspond to the reporting changes.
- Store a single `shared_pointer<UnwindRow>` as the current row, the analysis can read it, and it will be freed when no one has access to it. Aside from using smart pointers, I don't see a way to properly implement this.

Let me know which one I should pursue.

Commit:
Sync an assertion message with the refactors happened around it

Commit:
Merge branch 'main' into cfi-validation-generation

Commit:
Merge branch 'main' into cfi-validation-generation

Commit:
Sync implementation and `DWARFCFIAnalysis` explaination

Commit:
Change error messages for CFA

Commit:
Test CFA corner cases

statham-arm commented on llvm/include/llvm/DWARFCFIChecker/DWARFCFIState.h:
```diff
@@ -23,16 +23,23 @@ namespace llvm {
+/// way to modify it is to append a new row by updating it with a CFI directive,
+/// and the only way to read from it is to read the last row (i.e., current CFI
+/// state) from the table. The fetched row is constant and should not be
+/// modified or deleted.
```
I see; so you don't need to keep all the rows _right back to the start of the function_, only the ones between the `PrevRow` and `NextRow` rows in `DWARFCFIAnalysis::update`?

So it would be enough to give `DWARFCFIState` a method to _nearly_ empty its vector (reducing it to just one entry which was previously the final entry). Then `DWARFCFIAnalysis::update` could start by calling that. And now the size of `DWARFCFIState` doesn't grow in proportion to the size of the whole function, only the number of table rows generated by a single directive at a time.

statham-arm replied:
But whichever approach you take, if you're storing information in a vector which is not _currently_ used for anything, please write down in a comment why you're storing it at all, and what it's intended to be used for in future.

amsen20 replied:
Fixed.
The state now only stores the last row, and the other rows are copied instead of being shared with a smart pointer.
Sharing through a pointer might save one copy (which may be optimized away by the compiler), and that one extra copy is OK to have. Each row size is linear to the number of register bytes.

statham-arm commented on llvm/lib/DWARFCFIChecker/DWARFCFIAnalysis.cpp:
```diff
@@ -195,53 +200,59 @@ void DWARFCFIAnalysis::checkRegDiff(const MCInst &Inst, DWARFRegNum Reg,
-    }
-
+  // Each case is annotated with its corresponding number as described in the
+  // `DWARFCFIAnalysis`.
```
Could you add a reference to the specific _file_, please? "As described in `llvm/include/llvm/DWARFCFIChecker/DWARFCFIAnalysis.h`." If you're going to carefully map the code here to the comment in another file, it should be as easy as possible for someone to find that comment to compare against.

amsen20 replied:
Fixed, I guess this was the only place I referenced another file documentation. Let me know if there is another place that I can fix.

statham-arm commented on llvm/lib/DWARFCFIChecker/DWARFCFIAnalysis.cpp:
```diff
@@ -195,53 +200,59 @@ void DWARFCFIAnalysis::checkRegDiff(const MCInst &Inst, DWARFRegNum Reg,
+  // Case 2
+  if (PrevLoc.getLocation() != NextLoc.getLocation()) { // Case 2.a
+    Context->reportWarning(Inst.getLoc(),
+                           formatv("uncheckable change happened to register "
```
A small wording point: could you make sure this checker _takes the blame_ for not being able to check this?

The current wording is ambiguous enough that a user might think it's their own fault. "Is my change uncheckable? Should I have written a different one to make it easier to check?" But in fact many of these changes are checkable _in principle_, it's just that this particular checker doesn't have the ability to check them yet.

I think wording along the lines of "checking correctness of changes to register unwinding rules is not yet implemented" would communicate more clearly what the user needs to know:

1. it's not my fault
2. there's nothing I can change in my code to make this warning go away
3. the checker isn't telling me my rule is wrong, just that it doesn't know how to make sure it's right
4. maybe in a future version this will be better.

amsen20 replied:
Fixed

Commit:
Warn the user about ignoring `.cfi_escape` directive

dwblaikie commented on llvm/tools/llvm-mc/llvm-mc.cpp:
```diff
@@ -519,8 +526,15 @@ int main(int argc, char **argv) {
-  if (FileType == OFT_AssemblyFile) {
+  if (ValidateCFI) {
+    assert(FileType == OFT_Null);
+    auto *FFS = new CFIFunctionFrameStreamer(Ctx, std::move(FFA));
```
prefer make_unique here

amsen20 replied:
Fixed

Commit:
Move warning to the point happening

Commit:
Support `.cfi_label`

Commit:
Store only the last row in the state

Commit:
Reference another file properly

Commit:
Sort fields by size for memory layout and move them to the end of the class definition

Commit:
Change the error messages to show that the warnings are not the programmer's fault and the validation is not implemented yet.

Change the tests accordingly

Commit:
Delete unwanted commited test

Commit:
Use make unique instead of `new`ing a raw pointer

statham-arm reviewed (approved):
Thanks for all the fixes!

I've run out of things to complain about now :slightly_smiling_face: but I see that Petr has just nominated a couple more reviewers, so please give those people time to have thoughts as well.

Sterling-Augustine commented on llvm/lib/DWARFCFIChecker/DWARFCFIFunctionFrameStreamer.cpp:
```diff
@@ -0,0 +1,97 @@
+
+void CFIFunctionFrameStreamer::emitInstruction(const MCInst &Inst,
+                                               const MCSubtargetInfo &STI) {
+  updateReceiver();
```
This feels weird to me. updateReceiver works on LastInstruction, but not the instruction that came via emitInstruction. So updateReceiver defers analyzing Inst until the next call. It seems likely that there is a good reason for this, but I can't tell what it is.

At the very least this needs some kind of comment explaining what is going on. I might even have updateReciever take the instruction to analyze as a parameter, which makes things even more clear that current instruction is being deferred.

By my count, updateReciever has three call sites, and I can't tell off the top of my head which ones I would expect LastInstruction to be not present. I think two? Given that it is only used at the end of the function, perhaps it makes sense to split that portion into a new function that is called more selectively? (But perhaps that isn't possible.

In any event. Needs comments. Especially on why the very last LastInstruction will get analyzed even though it is deferred here.

dwblaikie commented on llvm/tools/llvm-mc/llvm-mc.cpp:
```diff
@@ -519,8 +526,15 @@ int main(int argc, char **argv) {
   std::unique_ptr<MCInstrInfo> MCII(TheTarget->createMCInstrInfo());
   assert(MCII && "Unable to create instruction info!");
 
+  auto FFA = std::make_unique<CFIFunctionFrameAnalyzer>(Ctx, *MCII);
```
Why is this constructed here, rather than created down at line 534 - the only place it's used?

amsen20 replied:
Moved.

dwblaikie commented on llvm/lib/DWARFCFIChecker/Registers.h:
```diff
@@ -0,0 +1,68 @@
+inline MCPhysReg getSuperReg(const MCRegisterInfo *MCRI, MCPhysReg Reg) {
+  if (isSuperReg(MCRI, Reg))
+    return Reg;
+  for (auto SuperReg : MCRI->superregs(Reg)) {
```
Some inconsistency with bracing in this patch - like this loop has braces, but a very similar loop at line 49 doesn't... I think technically according to the style guide, both should omit braces, but I could go either way.

(there's also some other rather long-bodied `for/if{}` without the braces on the for which get a bit of a stretch - but, yeah, the rules are pretty vague)

Commit:
Fix a typo

Commit:
[TEMP] Store FrameStreamer state in stack instead of trying to guess the previous state

Commit:
Fix the nested frame case and document the updateReceiver behavior

amsen20 commented on llvm/include/llvm/DWARFCFIChecker/DWARFCFIFunctionFrameAnalyzer.h:
```diff
@@ -0,0 +1,49 @@
+/// instantiates a `DWARFCFIAnalysis` for each frame. The errors/warnings are
+/// emitted through the `MCContext` instance to the constructor. If a frame
+/// finishes without being started or if all the frames are not finished before
+/// this classes is destructured, the program fails through an assertion.
```
Fixed

amsen20 replied:
fixed

Commit:
Move the definition to usage

Commit:
Sync the braces styles

amsen20 committed:
Merge branch 'main' into cfi-validation-generation

amsen20 committed:
Add triple setting to llvm-mc args to make sure that tests are running for the specific X86 arch

Commit:
Add overview for the link mentioned in `DWARFAnalysis`

amsen20 committed:
Update llvm/include/llvm/DWARFCFIChecker/DWARFCFIAnalysis.h

Co-authored-by: Paul Kirth <paulkirth@google.com>

Commit:
Lint

Commit:
Fix cases formatting

amsen20 commented on llvm/include/llvm/DWARFCFIChecker/DWARFCFIAnalysis.h:
```diff
@@ -0,0 +1,97 @@
+/// In each update, for each register (or CFA), the following cases can happen:
+/// 1. The unwinding rule is not changed:
+///   a. The registers involved in this rule are not modified: the analysis
+///   proceeds without emitting error or warning.
```
Fixed

Commit:
Use `SmallVector` instead of `std::vector`

amsen20 committed:
Update llvm/include/llvm/DWARFCFIChecker/DWARFCFIState.h

Co-authored-by: Paul Kirth <paulkirth@google.com>

Commit:
Lint

Commit:
Reword the future support for dwarf expressions

Commit:
Fix punctuation error

Commit:
Reformat Run tests to one line

amsen20 commented on llvm/test/DWARFCFIChecker/X86/spill-two-reg.s:
```diff
@@ -0,0 +1,57 @@
+        .cfi_same_value %rsi
+
+        pushq   %rbp
+        # CHECK: warning: CFA offset is changed from 8 to 16, and CFA register RSP is modified, but validating the modification amount is not implemented yet
```
Yes they do, should I mention them in the CHECKS?

Commit:
Change TODO style in all tests

amsen20 commented on llvm/test/DWARFCFIChecker/X86/nested-frames.s:
```diff
@@ -0,0 +1,31 @@
+.pushsection A
+f: 
+.cfi_startproc
+# TODO: Remove this line when the initial frame directives set the callee saved registers
```
Fixed

amsen20 commented on llvm/test/DWARFCFIChecker/X86/empty-section.s:
```diff
@@ -0,0 +1,11 @@
+# RUN: llvm-mc -triple x86_64-pc-linux-gnu %s --validate-cfi --filetype=null 2>&1 \
+# RUN:   | FileCheck %s --allow-empty 
```
I need it to check if my tool is actually emitting nothing. I cannot set the output as `asm` because the tool output is only errors and warnings, and when you are running the tool, it asserts that the output should be `null`.

amsen20 replied:
I can modify the tool to emit assembly code as well while validating, but in this PR, I tried to keep my changes to `llvm-mc.cpp` as minimal as possible. I guess this should be done in another PR if needed

Commit:
Check llvm-mc does not emit any error or warnings for each test

amsen20 commented on llvm/test/DWARFCFIChecker/X86/spill-two-reg.s:
```diff
@@ -0,0 +1,57 @@
+# RUN: llvm-mc -triple x86_64-pc-linux-gnu %s --validate-cfi --filetype=null 2>&1 \
+# RUN:   | FileCheck %s 
```
Added llvm-mc syntax check to all tests

Commit:
Replace `CHECK-NOT` with `implicit-check-not`

amsen20 commented on llvm/test/DWARFCFIChecker/X86/empty-section.s:
```diff
@@ -0,0 +1,11 @@
+# RUN: llvm-mc -triple x86_64-pc-linux-gnu %s --validate-cfi --filetype=null 2>&1 \
+# RUN:   | FileCheck %s --allow-empty 
+# CHECK-NOT: warning:
```
Fixed

Commit:
Mention that remember state is not supported

Commit:
Test all CFA directives

Commit:
Mention why remember and restore state is not supported

Commit:
Document why some directives are not supported and test them

Commit:
Merge branch 'main' into cfi-validation-generation

Commit:
Depends on DebugInfoDWARFLowLevel instead of DebugInfoDWARF

Commit:
Mention why --alow-empty is used for --validate-cfi tests and how it can be removed

Commit:
Include low level debug info library inside headers

amsen20 commented:
> lit.local.cfg

@dyung I changed the PR and added `lit.local.cfg` that ignores the target if it doesn't support X86, but I cannot test it on Mac by myself.
Do you have any idea how I can test if it fixes it or not?
